File: server/lib/face_crop.py
#!/usr/bin/env python3
"""Face-crop helper: reads an image, crops around the detected face at face_factor=0.8, writes result.

Usage:
    python face_crop.py <input_path> <output_path> [size] [face_factor]

Falls back to center-top crop if no face is detected.
"""
import sys
import os

def main():
    if len(sys.argv) < 3:
        print("Usage: face_crop.py <input> <output> [size] [face_factor]", file=sys.stderr)
        sys.exit(1)

    input_path = sys.argv[1]
    output_path = sys.argv[2]
    size = int(sys.argv[3]) if len(sys.argv) > 3 else 200
    face_factor = float(sys.argv[4]) if len(sys.argv) > 4 else 0.8

    from face_crop_plus import Cropper
    from PIL import Image
    import tempfile, shutil

    # face-crop-plus works on directories, so use temp dirs
    tmp_in = tempfile.mkdtemp()
    tmp_out = tempfile.mkdtemp()

    try:
        ext = os.path.splitext(input_path)[1] or '.jpg'

        def has_black_corners(img_path, threshold=10):
            """Check if the image has black (unfilled) corners."""
            img = Image.open(img_path)
            pixels = img.load()
            w, h = img.size
            corners = [(0, 0), (w-1, 0), (0, h-1), (w-1, h-1)]
            for x, y in corners:
                p = pixels[x, y]
                if isinstance(p, tuple):
                    if all(c < threshold for c in p[:3]):
                        return True
                elif p < threshold:
                    return True
            return False

        def try_crop(ff):
            """Run face_crop_plus with given face_factor, return output path or None."""
            # Clean temp dirs
            for f in os.listdir(tmp_in):
                os.remove(os.path.join(tmp_in, f))
            for f in os.listdir(tmp_out):
                os.remove(os.path.join(tmp_out, f))
            shutil.copy2(input_path, os.path.join(tmp_in, f"photo{ext}"))
            cropper = Cropper(face_factor=ff, output_size=size, strategy="largest")
            cropper.process_dir(input_dir=tmp_in, output_dir=tmp_out)
            results = [f for f in os.listdir(tmp_out) if not f.startswith('.')]
            if results:
                return os.path.join(tmp_out, results[0])
            return None

        # Try with requested face_factor, increase if black corners appear
        result_path = None
        used_ff = face_factor
        for ff in [face_factor, 0.5, 0.6, 0.7, 0.8]:
            result_path = try_crop(ff)
            if result_path and not has_black_corners(result_path):
                used_ff = ff
                break
            used_ff = ff

        if result_path:
            shutil.copy2(result_path, output_path)
            print(f"OK:face_detected:ff={used_ff}")
        else:
            # Fallback: center-top crop with PIL
            img = Image.open(input_path)
            w, h = img.size
            short = min(w, h)
            left = (w - short) // 2
            top = 0  # top-aligned for portraits
            img = img.crop((left, top, left + short, top + short))
            img = img.resize((size, size), Image.LANCZOS)
            img.save(output_path, "JPEG", quality=95)
            print("OK:fallback_crop")
    finally:
        shutil.rmtree(tmp_in, ignore_errors=True)
        shutil.rmtree(tmp_out, ignore_errors=True)

if __name__ == "__main__":
    main()

File: server/lib/face-crop.ts
/**
 * Face-crop helper: detects face in an image and crops around it.
 * Uses @vladmandic/face-api (TensorFlow.js) for detection + Sharp for cropping.
 * Falls back to center-top crop if no face is detected.
 */
import path from 'path';
import util from 'util';
import sharp from 'sharp';

// Polyfill for Node.js v25+ which removed util.isNullOrUndefined
// (required by @tensorflow/tfjs-node)
if (!(util as any).isNullOrUndefined) {
  (util as any).isNullOrUndefined = (v: any) => v === null || v === undefined;
}

let faceApiLoaded = false;
let faceapi: typeof import('@vladmandic/face-api');
let tf: any;

const MODEL_DIR = path.resolve(
  require.resolve('@vladmandic/face-api'),
  '../../model'
);

async function ensureModelsLoaded() {
  if (faceApiLoaded) return;

  tf = require('@tensorflow/tfjs-node');
  await tf.ready();

  faceapi = require('@vladmandic/face-api');

  // Load only the tiny face detector (fast + lightweight)
  await faceapi.nets.tinyFaceDetector.loadFromDisk(MODEL_DIR);
  faceApiLoaded = true;
}

interface FaceCropResult {
  method: 'face_detected' | 'fallback_crop';
}

/**
 * Crop an image around the detected face and save as JPEG.
 * @param inputPath  Path to the source image
 * @param outputPath Path for the cropped output
 * @param size       Output size in pixels (square)
 * @param faceFactor How much space around the face (0-1, higher = tighter crop)
 */
export async function faceCrop(
  inputPath: string,
  outputPath: string,
  size: number = 200,
  faceFactor: number = 0.4
): Promise<FaceCropResult> {
  await ensureModelsLoaded();

  // Read image metadata
  const metadata = await sharp(inputPath).metadata();
  const imgWidth = metadata.width!;
  const imgHeight = metadata.height!;

  // Decode image to raw RGB pixel buffer for face-api
  const { data, info } = await sharp(inputPath)
    .removeAlpha()
    .raw()
    .toBuffer({ resolveWithObject: true });

  // Create tensor from raw pixels
  const tensor = tf.tensor3d(
    new Uint8Array(data),
    [info.height, info.width, 3] // RGB
  );

  // Detect face
  const detection = await faceapi.detectSingleFace(
    tensor as any,
    new faceapi.TinyFaceDetectorOptions({ inputSize: 416, scoreThreshold: 0.3 })
  );

  tensor.dispose();

  if (detection) {
    const box = detection.box;

    // Expand the crop area around the face
    const faceWidth = box.width;
    const faceHeight = box.height;
    const faceCenterX = box.x + faceWidth / 2;
    const faceCenterY = box.y + faceHeight / 2;

    // Calculate crop size: face takes up `faceFactor` of the final crop
    const cropSize = Math.max(faceWidth, faceHeight) / faceFactor;

    // Center crop on the face, offset slightly upward (forehead room)
    const cropCenterY = faceCenterY - cropSize * 0.05;

    let left = Math.round(faceCenterX - cropSize / 2);
    let top = Math.round(cropCenterY - cropSize / 2);
    let cropW = Math.round(cropSize);
    let cropH = Math.round(cropSize);

    // Clamp to image bounds
    left = Math.max(0, Math.min(left, imgWidth - 1));
    top = Math.max(0, Math.min(top, imgHeight - 1));
    cropW = Math.min(cropW, imgWidth - left);
    cropH = Math.min(cropH, imgHeight - top);

    // Make it square
    const side = Math.min(cropW, cropH);
    cropW = side;
    cropH = side;

    await sharp(inputPath)
      .extract({ left, top, width: cropW, height: cropH })
      .resize(size, size)
      .jpeg({ quality: 95 })
      .toFile(outputPath);

    return { method: 'face_detected' };
  }

  // Fallback: center-top crop
  const short = Math.min(imgWidth, imgHeight);
  const left = Math.round((imgWidth - short) / 2);
  const top = 0; // top-aligned for portraits

  await sharp(inputPath)
    .extract({ left, top, width: short, height: short })
    .resize(size, size)
    .jpeg({ quality: 95 })
    .toFile(outputPath);

  return { method: 'fallback_crop' };
}

export { faceCrop as processPhoto };
